File: Programmers/Level1/실패율/실패율.js
function solution(N, stages) {
    let answer = new Map();
    let clear = stages.length;
    for (let i = 1; i < N + 1; i++) {
        let stop = stages.filter((e) => e === i).length;
        if (stop === 0) {
            answer.set(i, 0);
            continue;
        }
        answer.set(i, stop / clear);
        clear -= stop;
    }
    return [...answer].sort((a, b) => b[1] - a[1]).map((e) => e[0]);
}

// 다른 풀이
function solution(N, stages) {
    let answer = new Map();

    for (let i = 1; i < N + 1; i++) {
        let clear = stages.filter((e) => e >= i).length;
        let stop = stages.filter((e) => e === i).length;
        answer.set(i, stop / clear);
    }
    return [...answer].sort((a, b) => b[1] - a[1]).map((e) => e[0]);
}

File: Programmers/Level1/실패율/실패율.py
# dictionary 사용 코드
def solution(N, stages):
    answer = dict()
    clear = len(stages)
    
    stop = [stages.count(i) for i in range(1, N + 1)]
    
    for i in range(len(stop)):
        if stop[i] == 0:
            answer[i + 1] = 0
        else:
            answer[i + 1] = stop[i] / clear
            clear -= stop[i]
    
    return sorted(answer, key=lambda x:answer[x], reverse = True)

# list 사용 코드
def solution(N, stages):
    answer = []
    clear = len(stages)
    
    stop = [stages.count(i) for i in range(1, N + 1)]
    
    for i in range(len(stop)):
        if stop[i] == 0:
            answer.append([i + 1, 0])
        else:
            answer.append([i + 1, stop[i] / clear])
            clear -= stop[i]
    
    answer.sort(key=lambda x:(-x[1], x[0]))

    return [a[0] for a in answer]